### PR TITLE
[Woo POS] Architecture improvements: minor updates

### LINKED ISSUE #14405: [Woo POS] Consider re-adding property for `shouldShowClearCartButton` to aggregate model or `CartView`
> I'll keep it in mind and perhaps we can add it back at the end.... But for the time being I'm trying to get in the mindset of simply relying on the raw state we care about in the view as much as possible, without wrapping it up in extra functions and tests.
> 
> `posModel.cart.isNotEmpty && cartViewModel.canDeleteItemsFromCart` is already quite simple, and the `canDelete` bit is a wrapper around the current order stage (it needs to be `building`.) So I'm working towards it using those state items directly, and then we don't need to rely on combine flows to set flags, as we have here. 
> 
> That paves the way for us to try `@Observable` rather than `ObservableObject`, and has the potential to be more performant, as only changes to `orderStage` or `cart.isNotEmpty` would trigger this view being redrawn, rather than any change to `posModel` or `viewModel`, as happens with published properties on observable objects.

_Originally posted by @joshheald in https://github.com/woocommerce/woocommerce-ios/pull/14400#discussion_r1840086571_

### LINKED ISSUE #14563: [Woo POS] Improve Hashable and Equatable conformance for OrderSyncProductInput.ProductType
These need work, and tests. At the moment, they're only used for grouping the OrderItems correctly, using the dictionary functions, in `createGroupedOrderSyncProductInputs`. I'll tackle this in the variations project.

_Originally posted by @joshheald in https://github.com/woocommerce/woocommerce-ios/pull/14562#discussion_r1863724016_

### PR DESCRIPTION
**Pull request by joshheald**
Closes: #14405
Closes: #14563

## Description

This PR addresses a couple of minor changes left from previous PRs.

There is no change in behaviour. Unit tests should pass, and the POS should continue to work as it did before.

## Testing information

Areas to test:
Clear button should be available when viewing the item list with something in the cart. It should not be shown when the cart is empty, or when viewing checkout.

Editing the order should work as it did before: i.e. moving to the checkout without making any changes since the last time you viewed checkout should not result in the order being updated. If you go back and make changes, the next time you tap check out the order should be re-synced.

---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

## Reviewer (or Author, in the case of optional code reviews):

Please make sure these conditions are met before approving the PR, or request changes if the PR needs improvement:

- [x] The PR is small and has a clear, single focus, or a valid explanation is provided in the description. If needed, please request to split it into smaller PRs.
- [x] Ensure Adequate Unit Test Coverage: The changes are reasonably covered by unit tests or an explanation is provided in the PR description.
- [x] Manual Testing: The author listed all the tests they ran, including smoke tests when needed (e.g., for refactorings). The reviewer confirmed that the PR works as expected on all devices (phone/tablet) and no regressions are added.

### CODE REVIEW

**Comment by dangermattic** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">View files have been modified, but no screenshot or video is included in the pull request. Consider adding some for clarity.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: Yosemite/Yosemite/Stores/Order/OrderSyncProductInput.swift [hunk 1]
```diff
@@ -27,24 +38,26 @@ public struct OrderSyncProductInput {
...
     public enum ProductType: Hashable {
-        case product(OrderSyncProductTypeProtocol)
+        case product(any OrderSyncProductTypeProtocol)
         case variation(ProductVariation)
 
         public func hash(into hasher: inout Hasher) {
             switch self {
             case .product(let product):
-                hasher.combine("product-\(product.productID)")
+                hasher.combine("productType-product-")
```
Comment (jaclync) [inline Yosemite/Yosemite/Stores/Order/OrderSyncProductInput.swift]: super nit: feels like there's something after the last dash, maybe can remove it like in the variation case?

```suggestion
                hasher.combine("productType-product")
```

File: WooCommerce/Classes/POS/Presentation/CartView.swift [hunk 2]
```diff
@@ -147,6 +147,12 @@ private extension CartView {
             orderState: posModel.orderState,
             paymentState: posModel.paymentState)
     }
+
+    var shouldShowClearCartButton: Bool {
```
Comment (jaclync) [inline WooCommerce/Classes/POS/Presentation/CartView.swift:151]: When using `ObservableObject` before I used to worry about derived variables not triggering redraws when the source `@Published var`'s change, looks like this isn't a concern in the aggregated model architecture anymore.

**Review by jaclync — APPROVED** [review]
The Hashable changes took me some time to follow, anyway LGTM - testing worked as expected :shipit:

File: WooCommerce/Classes/POS/Presentation/CartView.swift (hunk 2 above)
Comment (joshheald) [reply WooCommerce/Classes/POS/Presentation/CartView.swift:151]: We'll have to see how that goes when we shift to `Observation`, which is more granular. It should still track access via computed vars, though the sessions talk more about computeds on the `@Observable`, not on the view.

For `ObservableObject`, you don't really have to worry. Any view which has a dependency (marked with `@ObservedObject`, `@StateObject`, or `@EnvironmentObject`) will be redrawn whenever any published property on that object changes.

File: Yosemite/Yosemite/Stores/Order/OrderSyncProductInput.swift (hunk 1 above)
Comment (joshheald) [reply Yosemite/Yosemite/Stores/Order/OrderSyncProductInput.swift]: Thanks, good spot (though I dread having to get this through CI again 🙄)

**Pull request merged**